src/dex/maverick-v2/maverick-v2-pool.ts

```diff
@@ -0,0 +1,279 @@
+import { Interface } from '@ethersproject/abi';
+import { DeepReadonly } from 'ts-essentials';
+import { Address, BlockHeader, Log, Logger, Token } from '../../types';
+import { catchParseLogError } from '../../utils';
+import { StatefulEventSubscriber } from '../../stateful-event-subscriber';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import { PoolState } from './types';
+import MaverickV2PoolABI from '../../abi/maverick-v2/MaverickV2Pool.json';
+import MaverickV2PoolLensABI from '../../abi/maverick-v2/MaverickV2PoolLens.json';
+import { AbiItem } from 'web3-utils';
+import { Contract } from 'web3-eth-contract';
+import { MaverickPoolMath } from './maverick-math/maverick-pool-math';
+import _ from 'lodash';
+
+export class MaverickV2EventPool extends StatefulEventSubscriber<PoolState> {
+  handlers: {
+    [event: string]: (
+      event: any,
+      state: DeepReadonly<PoolState>,
+      log: Readonly<Log>,
+      blockHeader: BlockHeader,
+    ) => DeepReadonly<PoolState> | null;
+  } = {};
+
+  poolContract: Contract;
+  poolLensContract: Contract;
+
+  logDecoder: (log: Log) => any;
+
+  addressesSubscribed: string[];
+
+  poolMath: MaverickPoolMath;
+
+  constructor(
+    readonly parentName: string,
+    protected network: number,
+    protected dexHelper: IDexHelper,
+    logger: Logger,
+    public tokenA: Token,
+    public tokenB: Token,
+    public feeAIn: bigint,
+    public feeBIn: bigint,
+    public tickSpacing: bigint,
+    public protocolFeeRatio: bigint,
+    public lookback: bigint,
+    public activeTick: bigint,
+    public address: Address,
+    public poolLensAddress: Address,
+    protected maverickV2Iface = new Interface(MaverickV2PoolABI),
+  ) {
+    const name = `${parentName.toLowerCase()}-${tokenA.symbol}-${
+      tokenB.symbol
+    }-${address.toLowerCase()}`;
+
+    super(parentName, name, dexHelper, logger);
+
+    // TODO: make logDecoder decode logs that
+    this.logDecoder = (log: Log) => this.maverickV2Iface.parseLog(log);
+    this.addressesSubscribed = [address];
+
+    // Add handlers
+    this.handlers['PoolAddLiquidity'] = this.handleAddLiquidityEvent.bind(this);
+    this.handlers['PoolRemoveLiquidity'] =
+      this.handleRemoveLiquidityEvent.bind(this);
+    this.handlers['PoolSwap'] = this.handleSwapEvent.bind(this);
+
+    this.poolMath = new MaverickPoolMath(
+      feeAIn,
+      feeBIn,
+      lookback,
+      tickSpacing,
+      protocolFeeRatio,
+      activeTick,
+      BigInt(tokenA.decimals),
+      BigInt(tokenB.decimals),
+    );
+
+    this.poolContract = new this.dexHelper.web3Provider.eth.Contract(
+      MaverickV2PoolABI as AbiItem[],
+      this.address,
+    );
+
+    this.poolLensContract = new this.dexHelper.web3Provider.eth.Contract(
```

Comment:
maybe we could reuse this contract for all instances instead of creating one for each pool

Reply:
fixed